File: scripts/build_gui.js
const fs = require('fs');
const path = require('path');

const FRONTEND_DIR = path.join(__dirname, '..', 'frontend');
const OUTPUT_FILE = path.join(__dirname, '..', 'src', 'UserControlGUI.cpp');

function readFile(filepath) {
    return fs.readFileSync(filepath, 'utf8');
}

function buildGui() {
    console.log("Iniciando build del frontend MDE...");
    
    // 1. Leer index.html
    let htmlContent = readFile(path.join(FRONTEND_DIR, 'index.html'));
    
    // 2. Leer e inyectar CSS
    const cssContent = readFile(path.join(FRONTEND_DIR, 'css', 'styles.css'));
    htmlContent = htmlContent.replace('<!-- INJECT_CSS -->', `<style>\n${cssContent}\n</style>`);
    
    // 3. Leer e inyectar JS
    const jsFiles = [
        'js/ui.js',
        'js/websocket.js',
        'js/joystick.js',
        'js/trajectory.js',
        'js/app.js'
    ];
    
    let jsCombined = "";
    for (const jsFile of jsFiles) {
        jsCombined += `\n// --- ${jsFile} ---\n`;
        jsCombined += readFile(path.join(FRONTEND_DIR, jsFile));
        jsCombined += "\n";
    }
        
    htmlContent = htmlContent.replace('<!-- INJECT_JS -->', `<script>\n${jsCombined}\n</script>`);
    
    // 4. Generar el código C++ final
    const cppTemplate = `#include "UserControlGUI.h"\n\nconst char* HTML_PAGE = R"EOF(\n${htmlContent}\n)EOF";\n\nconst char* get_html_page() {\n    return HTML_PAGE;\n}\n`;

    // 5. Escribir el archivo
    fs.writeFileSync(OUTPUT_FILE, cppTemplate, 'utf8');
        
    console.log(`Build exitoso: ${OUTPUT_FILE} generado correctamente.`);
}

buildGui();

File: scripts/build_gui.py
import os

FRONTEND_DIR = os.path.join(os.path.dirname(__file__), '..', 'frontend')
OUTPUT_FILE = os.path.join(os.path.dirname(__file__), '..', 'src', 'UserControlGUI.cpp')

def read_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        return f.read()

def build_gui():
    print("Iniciando build del frontend MDE...")
    
    # 1. Leer index.html
    html_content = read_file(os.path.join(FRONTEND_DIR, 'index.html'))
    
    # 2. Leer e inyectar CSS
    css_content = read_file(os.path.join(FRONTEND_DIR, 'css', 'styles.css'))
    html_content = html_content.replace('<!-- INJECT_CSS -->', f'<style>\n{css_content}\n</style>')
    
    # 3. Leer e inyectar JS
    # El orden es crítico para la Causa Eficiente (dependencias lógicas)
    js_files = [
        'js/ui.js',          # Objeto window.ui base
        'js/websocket.js',   # Comunicación
        'js/joystick.js',    # Componente de control
        'js/trajectory.js',  # Componente planificador
        'js/app.js'          # Orquestador final
    ]
    
    js_combined = ""
    for js_file in js_files:
        js_combined += f"\n// --- {js_file} ---\n"
        js_combined += read_file(os.path.join(FRONTEND_DIR, js_file))
        js_combined += "\n"
        
    html_content = html_content.replace('<!-- INJECT_JS -->', f'<script>\n{js_combined}\n</script>')
    
    # 4. Generar el código C++ final
    cpp_template = f"""#include "UserControlGUI.h"

const char* HTML_PAGE = R"EOF(
{html_content}
)EOF";

const char* get_html_page() {{
    return HTML_PAGE;
}}
"""

    # 5. Escribir el archivo
    with open(OUTPUT_FILE, 'w', encoding='utf-8', newline='\n') as f:
        f.write(cpp_template)
        
    print(f"Build exitoso: {OUTPUT_FILE} generado correctamente.")

if __name__ == '__main__':
    build_gui()
